src/main/fc/fc_msp.c

```diff
@@ -297,26 +303,71 @@ static void serializeDataflashReadReply(sbuf_t *dst, uint32_t address, const uin
         readLen = bytesRemainingInBuf;
     }
     // size will be lower than that requested if we reach end of volume
-    if (readLen > flashfsGetSize() - address) {
+    const uint32_t flashfsSize = flashfsGetSize();
+    if (readLen > flashfsSize - address) {
         // truncate the request
-        readLen = flashfsGetSize() - address;
+        readLen = flashfsSize - address;
     }
     sbufWriteU32(dst, address);
-    if (!useLegacyFormat) {
-        // new format supports variable read lengths
-        sbufWriteU16(dst, readLen);
-        sbufWriteU8(dst, 0); // placeholder for compression format
-    }
 
-    // bytesRead will equal readLen
-    const int bytesRead = flashfsReadAbs(address, sbufPtr(dst), readLen);
-    sbufAdvance(dst, bytesRead);
+    // legacy format does not support compression; first 4K bytes shouldn't be compressed as they contain a text header
+    const uint8_t compressionMethod = (address < 4096 || useLegacyFormat) ? NO_COMPRESSION : HUFFMAN;
```

Comment:
Gains from `address < 4096` are marginal and I'll probably remove it for simplicity.
What should be done is add another U8 parameter to `MSP_DATAFLASH_READ` allowing to explicitly state the readiness to get either compressed or non-compressed blocks, defaulting to non-compressed to not break compatibility with older configurator and allow downloading the same log with compression on/off to compare resulting files.